text/0000-create-stackify-repo.md

```diff
@@ -0,0 +1,159 @@
+      Path to directory with CA Certs to add to image
+  --mixin-labels string
+      Comma-separated list of mixins to add to labels
+  --packages string
+      Comma-separated list of packages to add to image
+  --user-id string
+      CNB User ID
+  --group-id string
+      CNB Group ID
+  --build
+      Add build image metadata (must specify either --build or --run)
+  --run
+      Add run image metadata (must specify either --build or --run)
+```
+
+This will be called under the hood by the `pack stack create` command, which will take a TOML config file. Pack will have to copy the stackify image onto each base image and then run stackify with the specified inputs for each image.
+
+This command can be invoked by passing in `--run-image`, `--build-image`, or both, depending on which images you would like to have created.
+
+`pack stack create my-stack --run-image <run-tag> --build-image <build-tag> -c /path/to/config.toml`
+
+Config file:
+```
+[stack]
+id = "(required)"
+maintainer = "(optional)"
+homepage = "(optional)"
+description = "(optional)"
+
+[stack.distro]
+name = "(optional)"
+version = "(optional)"
+
+[stack.metadata]
+# arbitrary keys and values
+
+[build]
+base-image = "paketobuildpacks/build:base"
+
+[[build.ca-certs]]
+uri = "/path/to/dir/with/build-certs"
+
+[[build.ca-certs]]
+uri = "file:///path/to/ca-cert.pem"
+
+[[build.ca-certs]]
+uri = "https://example.com/ca-cert.pem"
+
+[[build.packages]]
+name = "cowsay"
+
+[[build.packages]]
+name = "fortune"
+
+[run]
+base-image = "paketobuildpacks/run:base"
+
+[[run.ca-certs]]
+path = "/path/to/dir/with/run-certs"
+
+[[run.packages]]
+name = "cowsay"
+
+[[run.packages]]
+name = "rolldice"
+
+[[mixin-labels]]
+name = "build:set=build-utils"
+
+[[mixin-labels]]
+name = "run:set=run-utils"
+
+[[mixin-labels]]
+name = "set=shared-utils"
+
+[user]
+userID = 1000
+groupID = 1000
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+* Stackify must be run on the image being "stackify-ed" in order to add ca-certificates and packages.
+  * Pack has to run stackify within the image that is being modified.
+* If we do not do enough validation, users may be able to successfully use stackify with incompatible or invalid base images.
+* Some of the CNB metadata can be difficult to generate, particularly the mixins
+
+# Alternatives
+[alternatives]: #alternatives
+* Without this tool users can still use Dockerfiles to extend base images and add the CNB metadata.
+    * However, this requires them to be aware of all of the relevant metadata needed on the image and maintain a Dockerfile to continuously update their images. 
+
+# Prior Art
+[prior-art]: #prior-art
+N/A
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+* How much validation should be done on the images, if any?
+* It will be challenging to generate a list of common mixins between the build and run images. Is it ok that we will just have `build:<mixin>` and `run:<mixin>` labels and no shared mixins?
```

Comment:
If I'm not mistaken, this an implementation detail/challenge and not a limitation of the current proposed schema (`mixin-labels`). I'm not sure what the challenge is but it does seem to be a concern that may be addressed separately as long as the expectation (aka contract is adhered to).

Reply:
For stacks that are defined by the buildpacks project (i.e. `io.buildpacks.stacks.bionic`), stackify can generate most of the mixins automatically since for the bionic stack, mixins (other than set mixins) correspond to packages on the image. Therefore, users would only specifcy `mixin-labels` on a bionic stack if they want to add set mixins to the image metadata.

The challenge is that since stackify is run separately on each image, it can't itself know which mixins are on both images and which are on only one image. Therefore, it will not be able to generate mixin labels like `apt` on both images, but will instead have to put `build:apt` on the build image and `run:apt` on the run image.

Perhaps this is a concern that can be addressed separately, just wanted to clarify the issue.